File: lib/actions/redaction.ts
const SENSITIVE_KEY_PATTERN =
  /(?:password|passwd|secret|token|api[_-]?key|apikey|authorization|cookie|session|private[_-]?key|credential|access[_-]?token|refresh[_-]?token)/i;

export interface RedactionOptions {
  enabled?: boolean;
  maxDepth?: number;
  maxArrayItems?: number;
  maxStringLength?: number;
  maxBytes?: number;
}

const DEFAULTS: Required<RedactionOptions> = {
  enabled: true,
  maxDepth: 8,
  maxArrayItems: 50,
  maxStringLength: 4096,
  maxBytes: 32_768,
};

export function redactForTelemetry(value: unknown, options: RedactionOptions = {}): unknown {
  const config = { ...DEFAULTS, ...options };
  if (!config.enabled) return boundValue(value, config);

  const seen = new WeakSet<object>();
  return boundValue(redactValue(value, config, seen, 0), config);
}

function redactValue(
  value: unknown,
  config: Required<RedactionOptions>,
  seen: WeakSet<object>,
  depth: number
): unknown {
  if (value === null || typeof value === "boolean" || typeof value === "number") return value;
  if (typeof value === "bigint") return value.toString();
  if (typeof value === "string") return truncateControlSafe(value, config.maxStringLength);
  if (typeof value === "undefined") return null;
  if (typeof value === "function" || typeof value === "symbol") return `[${typeof value}]`;

  if (depth >= config.maxDepth) return "[MaxDepth]";
  if (seen.has(value)) return "[Circular]";
  seen.add(value);

  if (Array.isArray(value)) {
    try {
      return value
        .slice(0, config.maxArrayItems)
        .map((entry) => redactValue(entry, config, seen, depth + 1));
    } finally {
      seen.delete(value);
    }
  }

  try {
    const out: Record<string, unknown> = Object.create(null);
    let keys: string[];
    try {
      keys = Object.keys(value as Record<string, unknown>);
    } catch {
      return "[UnserializableObject]";
    }
    for (const key of keys) {
      const safeKey = truncateControlSafe(key, 256);
      if (SENSITIVE_KEY_PATTERN.test(key)) {
        out[safeKey] = "[REDACTED]";
        continue;
      }
      let entry: unknown;
      try {
        entry = (value as Record<string, unknown>)[key];
      } catch {
        out[safeKey] = "[UnserializableProperty]";
        continue;
      }
      out[safeKey] = redactValue(entry, config, seen, depth + 1);
    }
    return out;
  } finally {
    seen.delete(value);
  }
}

function boundValue(value: unknown, config: Required<RedactionOptions>): unknown {
  const json = safeStringify(value);
  if (byteLength(json) <= config.maxBytes) return value;
  return {
    truncated: true,
    reason: "payload_limit",
    original_bytes: byteLength(json),
    preview: truncateControlSafe(json, Math.min(config.maxStringLength, config.maxBytes)),
  };
}

export function safeStringify(value: unknown): string {
  try {
    return JSON.stringify(value) ?? "null";
  } catch {
    return JSON.stringify(redactForTelemetry(value, { maxDepth: 4, maxBytes: 4096 })) ?? "null";
  }
}

function truncateControlSafe(value: string, maxLength: number): string {
  const sanitized = value.replace(/[\u0000-\u0008\u000B\u000C\u000E-\u001F\u007F]/g, "\uFFFD");
  return sanitized.length > maxLength ? `${sanitized.slice(0, maxLength)}...[truncated]` : sanitized;
}

function byteLength(value: string): number {
  return new TextEncoder().encode(value).length;
}

File: integrations/hermes/aurel_hermes/redaction.py
from __future__ import annotations

import json
import re
from typing import Any

SENSITIVE_KEY_PATTERN = re.compile(
    r"(password|passwd|secret|token|api[_-]?key|apikey|authorization|cookie|session|private[_-]?key|credential|access[_-]?token|refresh[_-]?token)",
    re.IGNORECASE,
)


def redact(value: Any, *, enabled: bool = True, max_payload_bytes: int = 32768) -> Any:
    if not enabled:
        return _bound(value, max_payload_bytes)
    return _bound(_redact_value(value, set(), 0), max_payload_bytes)


def _redact_value(value: Any, seen: set[int], depth: int) -> Any:
    if value is None or isinstance(value, (bool, int, float)):
        return value
    if isinstance(value, str):
        return _clean(value, 4096)
    if depth > 8:
        return "[MaxDepth]"

    obj_id = id(value)
    if isinstance(value, (dict, list, tuple, set)):
        if obj_id in seen:
            return "[Circular]"
        seen.add(obj_id)

    if isinstance(value, dict):
        try:
            out: dict[str, Any] = {}
            for key in list(value)[:100]:
                safe_key = _clean(str(key), 256)
                if SENSITIVE_KEY_PATTERN.search(str(key)):
                    out[safe_key] = "[REDACTED]"
                    continue
                try:
                    item = value[key]
                except Exception:
                    out[safe_key] = "[UnserializableProperty]"
                    continue
                out[safe_key] = _redact_value(item, seen, depth + 1)
            return out
        finally:
            seen.remove(obj_id)
    if isinstance(value, (list, tuple, set)):
        try:
            return [_redact_value(item, seen, depth + 1) for item in list(value)[:50]]
        finally:
            seen.remove(obj_id)
    return _clean(repr(value), 1024)


def _bound(value: Any, max_payload_bytes: int) -> Any:
    encoded = safe_json(value).encode("utf-8", errors="replace")
    if len(encoded) <= max_payload_bytes:
        return value
    preview = encoded[: min(max_payload_bytes, 4096)].decode("utf-8", errors="replace")
    return {"truncated": True, "reason": "payload_limit", "preview": preview}


def safe_json(value: Any) -> str:
    try:
        return json.dumps(value, separators=(",", ":"), ensure_ascii=False)
    except (TypeError, ValueError):
        return '"[Unserializable]"'


def _clean(value: str, max_len: int) -> str:
    cleaned = "".join(ch if ch >= " " or ch in "\n\r\t" else "\ufffd" for ch in value)
    return cleaned if len(cleaned) <= max_len else cleaned[:max_len] + "...[truncated]"
